### PR TITLE
DebugVertex: fix ShaderObject debugging

### PR DESCRIPTION
**Pull request by DHrpcs3**
m_SPIRV was empty because of wrong ShaderModule

### CODE REVIEW

File: renderdoc/driver/shaders/spirv/spirv_disassemble.cpp [hunk 1]
```diff
@@ -297,7 +297,8 @@ rdcstr Reflector::Disassemble(const rdcstr &entryPoint,
       "Generator: %s\n"
       "Generator Version: %u\n"
       "\n",
-      m_MajorVersion, m_MinorVersion, m_SPIRV[3], ToStr(m_Generator).c_str(), m_GeneratorVersion);
+      m_MajorVersion, m_MinorVersion, m_SPIRV.size() > 3 ? m_SPIRV[3] : 0,
```
Comment (baldurk) [inline renderdoc/driver/shaders/spirv/spirv_disassemble.cpp]: What's the purpose behind this change, to prevent crashes if invalid SPIR-V is passed in? This should really be checked for elsewhere and earlier as I wouldn't want to have checks for `m_SPIRV` throughout this code.

**Review by baldurk — CHANGES_REQUESTED** [review]
Thanks for this fix, it looks fine but from quickly checking I think the same fix needs to be applied to the other debug functions.

File: renderdoc/driver/shaders/spirv/spirv_disassemble.cpp (hunk 1 above)
Comment (DHrpcs3) [reply renderdoc/driver/shaders/spirv/spirv_disassemble.cpp]: It is not checked, that is why there segfault with latest release when you try to debug shader object.

```c++
shadRefl.PopulateDisassembly(shader->spirv);
```
is invoked unconditionally (almost, `shadRefl.refl->debugInfo.debuggable` is true for some reason, probably it is also bug), it contains following code:
```c++
if(disassembly.empty())
    disassembly = spirv.Disassemble(refl->entryPoint, instructionLines);
```
disassembly is empty, so it reaches to spirv.Disassemble and there m_SPIRV dereference without any checks

Comment (baldurk) [reply renderdoc/driver/shaders/spirv/spirv_disassemble.cpp]: I understand but is that only as a result of the earlier problem that you've fixed where an invalid/unbound shader module is referenced when using the shader object extension? with your fix, does this crash still happen and if so from where?

As I say this should be fixed elsewhere if it's a problem - we should never get as far as trying to disassemble with an invalid shader module so it suggests the root problem is earlier. The crash you're describing is the symptom, not the problem

Comment (DHrpcs3) [reply renderdoc/driver/shaders/spirv/spirv_disassemble.cpp]: > with your fix, does this crash still happen and if so from where?

No, there without surprises, it is dead code with fix

> As I say this should be fixed elsewhere if it's a problem - we should never get as far as trying to disassemble with an invalid shader module so it suggests the root problem is earlier. The crash you're describing is the symptom, not the problem

OK, if you think that 0 bounds is not good enough to make error visible, let me revert this change

**Comment by DHrpcs3** [comment]
> I think the same fix needs to be applied to the other debug functions.

If I will need to use other shader debug functions and I will have fix for them, I will make PR for sure

**Comment by baldurk** [comment]
That's fine if you don't want to iterate on this PR, but in that case can you please close this and file an issue so it can be tracked for a full fix? As a reproducible test case it sounds like you could provide any shader object demo such as Sascha Willems' rather than yours specifically.

**Comment by DHrpcs3** [comment]
What do you mean with don't want to iterate on this PR? Scope of this PR is fix debugging of vertex shader objects, not all shaders, not all issues in this repository. If it is not enough, sure I can close this PR or you can close it for me please.

Segfault can be reproduced by starting debugging of any vertex shader object which produces vertices, just click on any vertex in Mesh View and select Debug Vertex. You do not need to have special code, any example with shader objects should be enough.

**Comment by baldurk** [comment]
OK then, that's fine. I would appreciate if you could open an issue for this bug so that it can be properly tracked and a full fix made.
